### ISSUE TITLE
Deploy Review Apps

### ISSUE THREAD

**Issue by joaqtor** [rancher/rancher#6846]
Hi.

I'm using GitLab for CI/CD. Already have my staging environment working tidy with Rancher.

Last week, GitLab announced [Review Apps](https://about.gitlab.com/2016/11/22/introducing-review-apps/), which is basicaly a test environment for every merge request.

So if I open a merge request for the 22-header-color branch, the goal is to have instantly an endpoint in 22-header-color.mydomain.com with deployed branch image, for immediatly testing.

My question is how to set up this in Rancher, using Rancher's API or rancher-compose (I prefer API). I need a load balancer that dynamically points to the correct service given the requested subdomain? How can I automate this?

I would appreciate any help.

Thanks in advance.

**Comment by hwinkel** [comment]
does not only need the service named after the branch name

**Comment by joaqtor** [comment]
So I just need to name the service `22-header-color` and the routing will work out of the box?

**Comment by darrenscerri** [comment]
We have managed something similar mainly using Rancher CLI. Our CI/CD server builds docker images and automatically provisions our whole application stack with dynamic ports for that build. We don't use subdomains at the moment. We have created a "discovery" service that returns all exposed ports for each service which are then fed to our automated test suites.

We are quite satisfied with Rancher CLI:

Pseudo-code is as follows:

- create a docker-compose.yml that defines the app stack with all the required images and image tags
- `STACK_ID=rancher stack create` to create the stack on Rancher
- `rancher wait ${STACK_ID}` to wait for the stack to be ready
- `DISCOVERY_URL=$(rancher inspect $(rancher inspect ${STACK_ID} --format {{.name}})/discovery --format "{{(index (.publicEndpoints) 0).ipAddress}}:{{(index (.publicEndpoints) 0).port}}")` to get the exposed ip and port of our discovery service
- `curl ${DISCOVERY_URL} > urls.json` to save all URLs in a JSON file so that they are consumed by other services
- `rancher rm -s ${STACK_ID}` once you want to remove the stack
